### PR TITLE
Improve odsp driver fluid cache perf

### PR DESCRIPTION
**Pull request by jatgarg**
## Description
Item: https://dev.azure.com/fluidframework/internal/_workitems/edit/4040
Improve odsp driver fluid cache performance. Avoid closing db after each access and rather schedule close so that it can be used many times in between closes.
Schedule idle task after 2 sec for first time.

### CODE REVIEW

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (NicholasCouri) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:152]: out of curiosity, can the db close immediately after we getFluidCacheIndexedDbInstance ? Ex. event close ? If so, is it possible db can return undefined ?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (NicholasCouri) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:142]: >30000 [](http://example.com/codeflow?start=7&length=5)

why 30 secs ? Is it random ?

Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: Yes, random for now. Just want to keep it open for sometime. Should be enough for load/catching up etc.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:152]: What do you mean by db return undefined? Do you mean that openDB, function return undefined? It should not close automatically and if it does, then we have try/catch in functions using the db and we will just catch the expection and return undefined.

**Comment by christiango** [comment]
Can you please get a review from @JasonGore since we had a bunch of issues where open db connections were blocking signout

**Comment by christiango** [comment]
Have we done any performance tests here to validate this improves things? We have an indexeddb test app that my team developed that I can add you to. Would be great to see if this has a tangible impact in isolation before we add this complexity.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (christiango) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: This may cause issues for the host app, in particular we've seen this mean that signout hangs.

**Comment by jatgarg** [comment]
> Have we done any performance tests here to validate this improves things? We have an indexeddb test app that my team developed that I can add you to. Would be great to see if this has a tangible impact in isolation before we add this complexity.

Yes, you can add me to that. If you look at this item: https://dev.azure.com/fluidframework/internal/_workitems/edit/4040, in telemetry it says that we spent about 150-200 ms average in opening the db. So we are targeting that cost.

**Comment by christiango** [comment]
Can we make this configurable so the host can run this as an A/B test?

**Comment by JasonGore** [comment]
@christiango is right, leaving the DB open will prevent both the existing browser tab and other tabs from deleting the DB and clearing storage on signout, effectively hanging signout. I had to resolve this issue [with a PR](https://github.com/microsoft/FluidFramework/pull/14315) to ensure that DBs are closed after transactions are completed.

30 seconds is too long to wait for DB deletion to be unblocked, and this is assuming further DB transactions aren't started that reset the 30 second timer. This change should not be added as default behavior without making sure that all DB sessions have `versionchange` listeners that can abort transactions in order to unblock DB deletion. This is, however, a fairly invasive change that should have a listener pattern and/or enforcement and unit tests to ensure deletions are not blocked.

Can this behavior be made opt-in while preserving the existing close behavior?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: There are two possible solutions here I feel like:
- decrease timeout substantially. Even 1-2 seconds would result in over 99% of cases reusing opened DB. I'd start with 2 seconds, but also record if db was reused or not on each cache call.
- Not use timeouts but bake a notion of "session" into cache layer, such that driver can control precisely duration of session (when to close DB). It's a bit hard to implement it that way, as driver does not fully control whole load sequence - Loader does, but I think we need to make certain changes in driver interface (irrespective of that change) that will make it possible in the future.

One thing to add - it might be great to have explicit "release all resources" call on cache. Driver can use it when driver is disposed (for example error is hit). But driver could also call this API in future where we know we are don with load sequence.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:125]: This could have a race condition - what if two callers call into openDB at the same time?
I think you should do something liek that (unless you do not what to deal with Deferred objects or Lazy promises):

```ts
const db = await getFluidCacheIndexedDbInstance(this.logger);
if (this.db == undefined) {
    this.db = db;
} else {
    db.close();
}
```

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:137]: other than the race condition I mentioned above, sounds like this.dbCloseTimer should be always undefined here, no?
If that's true, maybe assert is better? Not sure.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (JasonGore) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: I was thinking that we should have some kind of DB session pattern that is required for all DB interactions to ensure that the session reacts to DB deletions (`versionchange`) events appropriately and backed with enforcement or unit tests to make sure the driver is not blocking DB deletions. I'm not sure if the notion of "session" you're describing overlaps or is higher in the hierarchy, but I'm thinking either approach would help us to ensure DB sessions behave consistently.

File: packages/drivers/driver-web-cache/src/test/FluidCache.test.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/test/FluidCache.test.ts:262]: we are not testing cases like timer expiration, "versionchange". These are the new states, so if feels like we need to test them. I'd at minimum add timer expiration test.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (JasonGore) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:130]: Sorry, I did miss this change with my earlier comments. I still think it would be good to guarantee this would apply to all cases with unit tests/patterns/etc. if there is no guarantee of that already.

Also good to note this will cause all existing DB sessions to throw errors, which could impact FF and the hosting app. Do you know what FF and app behavior will be in these cases?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: I'm not sure I follow comments about all sessions or impact on deletion. Could you please clarify that?
Before & after change have same behavior - DB is left opened for a while (i.e. code is async between open and close).
What is changed here is that duration of DB being opened is longer. Is this the primary concern? I.e. are we saying that keeping DB opened longer is bad because it will cause more trouble, and count wise we can go over some thresholds?
Or something else?
(Let's assume we will use 1-2 seconds, not 30 seconds for this discussion)

Comment (vladsud) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: RE session: I meant container load session. I.e. we know that we need to make couple calls to cache during load.
The problem is that driver does not control it today, it's controlled by Loader layer and it would be weird hard to add notion of session here.

Comment (JasonGore) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: > Before & after change have same behavior - DB is left opened for a while (i.e. code is async between open and close).

I don't think this is true for the before case, this is precisely what [the previous PR](https://github.com/microsoft/FluidFramework/pull/14315/files) I created changed, ensuring DBs are closed after they are open and used in a synchronous fashion. This is because DBs were left open and were blocking DB deletion in other functions and browser tabs. This PR seems to fundamentally change that behavior, reverting in part to leaving DBs open for some amount of time after they are used.

Comment (vladsud) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: There is await in between open & close. Even if it takes 100ms, other activity could be scheduled that hit exactly same condition - DB is opened and in use. I do not see a difference here, other than amount of time we are talking about,

Comment (JasonGore) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:142]: Sorry, yes, you are correct. The bigger problem before my PR is that there were code paths that would never close the DB, so yes, the amount of time is the biggest concern (and ensuring we don't add any new code paths that would never close the DB as was the case previously).

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (JasonGore) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:161]: Out of curiosity what does `assert` do in consumer code? I don't see any other instances of this in `FluidCache.ts`

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:82]: We are defaulting to closing the db immediately.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:164]: It will throw exception in case it is false. We use it in FF code.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (JasonGore) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:82]: Yeah, I think `immediateClose ?? true` may be better

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (JasonGore) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:164]: In that case, my next question is whether this is so exceptional that the module is unusable and worth blowing up the caller over? Or is it something that can be observed over telemetry but should otherwise work fine?

Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:164]: We don't blow up. All the db open are wrapped in try catches where we just record telemetry and go on.

**Comment by jatgarg** [comment]
@vladsud @NicholasCouri  ping

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:62]: I think you need to mention in comments, or in name of property what number means - is it in seconds, milliseconds, or something else?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:57]: Can we have same support by saying that closeDbAfter <=0 means immediate closure?
Maybe we want to keep zero as non-special (it's delayed closed as 0ms timer has lower priority than immediate work or micro-task). Not sure - feels like we do not care.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:83]: Should we try to enable delayed closing by default?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:57]: The reason to have one property - it's a bit ambiguous when using true / 2000 setting.

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (vladsud) [inline packages/drivers/driver-web-cache/src/FluidCache.ts:148]: It would be great to capture somehow in telemetry how often reuse happens vs. not

File: pnpm-lock.yaml
Comment (vladsud) [inline pnpm-lock.yaml:18197]: not sure if these changes are intentional

**Review by vladsud — APPROVED** [review]

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:83]: @christiango do you have any objection to enable it by default?

File: packages/drivers/driver-web-cache/src/FluidCache.ts
Comment (jatgarg) [reply packages/drivers/driver-web-cache/src/FluidCache.ts:146]: Currently recorded with FluidCacheAccess event only rather than introducing another event.

**Pull request merged**
